Documented the workflow metadata [redux]

pbowyer opened the pull request:
With @javiereguiluz's permission I have taken #9476 and incorporated the comments and code examples into it.

This is the start of a push to improve the Workflow documentation. The focus now is on expanding the documentation by merging pull requests; organizing it will come later.

javiereguiluz committed:
Documented the workflow metadata

pbowyer committed:
Add missing metadata key, fixing comment by @noniagriconomie on https://github.com/symfony/symfony-docs/pull/9476

pbowyer committed:
Incorporate the code examples in https://github.com/symfony/symfony-docs/pull/9476#discussion_r184728373 into the documentation

pbowyer commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+        $workflow = $registry->get($article);
+
+        // Or, if you don't inject the Workflow Registry, fetch from the Container:
+        $workflow = $this->get('workflow.article');
```
Do we want to encourage users to use the Container directly; or push them towards the registry (as I've used in the lines above)?

HeahDude replied:
In Symony 4.x versions we prefer to use action injection when it makes in controllers (or subscribed services).
It's ok here to type hint the registry in the action directly since it's a requirement for it.

You can remove those two lines thanks!

pbowyer commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+        ;
+    }
+
+There is a shortcut that works with everything::
```
I listed this separately, thinking there is a drawback to people using it (more magic/understanding required).

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+        framework:
+            workflows:
+                blog_publishing:
+                    metadata: 'Blog Publishing Workflow'
```
Should be:
```yaml
metadata:
    title: 'Blog Publishing Workflow'
```

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+        >
+
+            <framework:config>
+                <framework:workflow name="blog_publishing" type="workflow">
```
You should either remove the `type` here, or add it in other formats (php and yaml) too to remain consistent across examples with different config types.

pbowyer replied:
I've removed to make it consistent. There's a discussion to be had around whether we should be explicit about options, or exclude the defaults?

I have not found all possible options documented anywhere, and it would save developers time if they were listed.

HeahDude replied:
We encourage doing the following:
```bash
bin/console config:dump-reference framework workflows
```
Do you think we should make it more explicit somewhere in the docs of the integration of the component?

pbowyer replied:
> Do you think we should make it more explicit somewhere in the docs of the integration of the component?

Definitely! I didn't know that existed, and I've been using Symfony since 2.4...

pbowyer replied:
Added to the documentation in https://github.com/symfony/symfony-docs/pull/11209/commits/60aa8b16342fc1b798a307a4c038bbe148917fd5

HeahDude replied:
Thanks!

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+        $workflow = $this->get('workflow.article');
+
+        $workflow
+            ->getMetadataStore()
```
Code style to discuss, but I think this should be on the line above since the workflow is not being fluent here:
```php
$workflow->getMetadataStore()
    ->getWorkflowMetadata()['title'] ?? false
;
```

pbowyer replied:
I'll leave this change until the code style discussion has been had!

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+
+            // $transition = ...; (an instance of Transition)
+            $title = $this->get('workflow.article')->getMetadataStore()->getMetadata('title', $transition);
+            $request->getSession()->getFlashBag()->add('info', "You have successfully applied the transition with title: '$title'");
```
```diff
- $request->getSession()->getFlashBag()->add
+ $this->addFlash
```

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+    use Symfony\Component\Workflow\Event\GuardEvent;
+    use Symfony\Component\Workflow\TransitionBlocker;
+
+    class DoneGuard implements EventSubscriberInterface
```
`OverdueGuard`?

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+            $title = $this->get('workflow.article')->getMetadataStore()->getMetadata('title', $transition);
+            $request->getSession()->getFlashBag()->add('info', "You have successfully applied the transition with title: '$title'");
+
+In a listener, access via the Event::
```
Maybe we could add a little introduction for this very nice example.

Some context about what it is doing and why it's useful to access metadata in there, also that we use a variant here `Event::getMetadata()`. What do you think?

pbowyer replied:
Those are all very good ideas 👍

pbowyer replied:
Added in https://github.com/symfony/symfony-docs/pull/11209/commits/0affc6b8a37ccf7ffcb141d960f3d20aa187766c.

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +450,216 @@ The following example shows these functions in action:
+
+In Twig templates, metadata is available via the ``workflow_metadata()`` function:
+
+.. code-block:: twig
```
Should be `html+twig`

HeahDude reviewed (commented):
This a great contribution, thanks a lot!

pbowyer committed:
First set of tidy-ups for @HeahDude's feedback.

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +450,215 @@ The following example shows these functions in action:
+Storing Metadata
+----------------
+
+.. versionadded:: 4.1
```
Please add a blank line after the directive

pbowyer replied:
Thanks @OskarStark, all included.

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +450,215 @@ The following example shows these functions in action:
+             ),
+         ));
+
+Then, you can access this metadata in your PHP code as follows:
```
“Then you can access this metadata in your controller as follows::” 

And remove the next sentence please

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +450,215 @@ The following example shows these functions in action:
+
+In a listener, access via the Event::
+
+    <?php
```
Please remove the php open tag

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +450,215 @@ The following example shows these functions in action:
+In a Flash message in your Controller::
+
+            // $transition = ...; (an instance of Transition)
+            // $workflow is a WorkFlow instance retrieved from the Registry (see above)
```
Please use “Workflow” instead of “WorkFlow”

pbowyer committed:
Incorporate @OskarStark's feedback

pbowyer committed:
Document how to see all configuration options (see https://github.com/symfony/symfony-docs/pull/11209#discussion_r268392587)

pbowyer committed:
Add an introduction as suggested in https://github.com/symfony/symfony-docs/pull/11209#discussion_r268391900

When https://github.com/symfony/symfony-docs/issues/9465 is finished, link to details on Transition Blockers.

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +459,220 @@ The following example shows these functions in action:
+
+The example below uses a new feature introduced in 4.1 called Transition Blockers. These let you
+return a user-friendly error message when you stop a transition from happening. In the example we
+get this user-friendly message from the Event's metadata, giving you an easy place to manage the
```
`get this message` would be enough the second time?

```diff
- Event
+ :class:`Symfony\\Component\\Workflow\\Event\\Event`
```

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +459,220 @@ The following example shows these functions in action:
+The example below uses a new feature introduced in 4.1 called Transition Blockers. These let you
+return a user-friendly error message when you stop a transition from happening. In the example we
+get this user-friendly message from the Event's metadata, giving you an easy place to manage the
+text. This is a contrived example; in production systems you may prefer to use the
```
"unreal", "hypothetic", "fictive" ... ? may be more friendly than "contrived". Personally I'm glad to have learned that word, thanks :D.

pbowyer replied:
Contrived is the most accurate word out of the four. Hypothetical doesn't have quite the same meaning.

javiereguiluz replied:
I'm afraid we can't use "contrived" because it's a complex English word (lots of non native speakers don't understand it, including myself). Please, replace it by any of these synonyms: https://www.thesaurus.com/browse/contrived Thanks!

HeahDude replied:
"fake" sounds good.

pbowyer replied:
> "fake" sounds good.

It's not a fake example: it's a real example (but one people should not emulate :) ). The synonyms are not accurate for this meaning (they are poor synonyms), so this must be totally rephrased.

Elsewhere in the documentation you must point out that an example is simplified and should not be used in production. Can you point me to the wording used?

pbowyer replied:
Resolving here to carry on in the new comment below

HeahDude commented on workflow/usage.rst:
```diff
@@ -450,3 +459,220 @@ The following example shows these functions in action:
+
+Metadata can also be accessed in a Listener, from the Event object.
+
+The example below uses a new feature introduced in 4.1 called Transition Blockers. These let you
```
You could say here "Using transition blockers you can return user-friendly...", and add a note below the example like:

```rst
.. versionadded:: 4.1

    The transition blockers were added in version 4.1.
```

pbowyer replied:
I've also created another PR to document transition blockers: https://github.com/symfony/symfony-docs/pull/11212

HeahDude reviewed (commented):
I've left some minor comments, but that looks great. Thank you again for your work on this.

Note that we can do some minor changes when merging if you don't want to bother with specific formatting.

pbowyer commented on workflow/usage.rst:
```diff
@@ -450,3 +459,220 @@ The following example shows these functions in action:
+            $title = $workflow->getMetadataStore()->getMetadata('title', $transition);
+            $this->addFlash('info', "You have successfully applied the transition with title: '$title'");
+
+In a listener, access via the Event
```
Thank you, removed

pbowyer commented:
@HeahDude re "specific formatting" I've noticed the file uses a mix of `array( ... )` and `[ ... ]`. Which should be used? 

When I took Javier's commits the branch ended up being created against `symfony-docs/master`. Can you make it be against the right branch when you merge?

HeahDude commented:
>I've noticed the file uses a mix of array( ... ) and [ ... ]. Which should be used?

Now we use the new short syntax, since the lowest supported version is 3.4 having PHP 5.5 as minimum.

>When I took Javier's commits the branch ended up being created against symfony-docs/master. Can you make it be against the right branch when you merge?

4.1 was master back then, now we need to merge into 4.2 but we can do it on merge thanks to the CLI tool we use to maintain the docs.
Thanks!

pbowyer committed:
Incorporate further excellent feedback from @HeahDude

pbowyer committed:
Update arrays to use short syntax

javiereguiluz commented on workflow/usage.rst:
```diff
@@ -450,3 +459,220 @@ The following example shows these functions in action:
+Using transition blockers you can
+return a user-friendly error message when you stop a transition from happening. In the example we
+get this message from the :class:`Symfony\\Component\\Workflow\\Event\\Event`'s metadata, giving
+you an easy place to manage the text. This is a contrived example; in production code you may
```
I think we can remove the entire phrase that starts with *"This is a contrived example..."* and add instead a comment like this one just before the `$explanation = ...` code:

```
// inject the 'translation' service if you need to translate the explanation
```

pbowyer replied:
I've simplified the english and turned it into a tip, as it's an aisde rather than part of the narrative.

pbowyer committed:
Simplify the English and turn it into a tip

OskarStark commented on workflow/usage.rst:
```diff
@@ -610,8 +610,12 @@ Metadata can also be accessed in a Listener, from the Event object.
+
+.. tip::
+
+This is a simple example; in production you may prefer to use the :doc:`Translation </components/translation>`
```
This is the Tip, so it need indention like a code block for example

pbowyer replied:
Thanks - wondered why it wasn't rendering properly

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+
+        // config/packages/workflow.php
+
+$container->loadFromExtension('framework', [
```
The indention is wrong, it needs indention like the php file path 2 lines above

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+Using transition blockers you can
+return a user-friendly error message when you stop a transition from happening. In the example we
+get this message from the :class:`Symfony\\Component\\Workflow\\Event\\Event`'s metadata, giving
+you an easy place to manage the text.
```
Please try to get rid of the „easy“

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+.. versionadded:: 4.1
+
+    The transition blockers were added in version 4.1.
+
```
„The transition blockers were introduced in Symfony 4.1.“

pbowyer committed:
Incorporate @OskarStark's feedback

pbowyer commented:
Re the code style comment above which is outstanding: Is there a linter for code examples? An automated formatter? 

I imagine we stick to Symfony's code style, but as Phpstorm's Symfony2 preset seems to differ in small ways from the current Symfony format, it would be good to have a definitive formatter to use.

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:framework="http://symfony.com/schema/dic/symfony"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
```
```suggestion
            xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+    .. code-block:: xml
+
+        <!-- config/packages/workflow.xml -->
+        <?xml version="1.0" encoding="utf-8" ?>
```
```suggestion
        <?xml version="1.0" encoding="UTF-8" ?>
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xmlns:framework="http://symfony.com/schema/dic/symfony"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd
+                http://symfony.com/schema/dic/symfony http://symfony.com/schema/dic/symfony/symfony-1.0.xsd"
```
```suggestion
                http://symfony.com/schema/dic/symfony https://symfony.com/schema/dic/symfony/symfony-1.0.xsd"
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+
+In a Flash message in your Controller::
+
+            // $transition = ...; (an instance of Transition)
```
```suggestion
    // $transition = ...; (an instance of Transition)

```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+In a Flash message in your Controller::
+
+            // $transition = ...; (an instance of Transition)
+            // $workflow is a Workflow instance retrieved from the Registry (see above)
```
```suggestion
    // $workflow is a Workflow instance retrieved from the Registry (see above)
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+
+            // $transition = ...; (an instance of Transition)
+            // $workflow is a Workflow instance retrieved from the Registry (see above)
+            $title = $workflow->getMetadataStore()->getMetadata('title', $transition);
```
```suggestion
    $title = $workflow->getMetadataStore()->getMetadata('title', $transition);
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+            // $transition = ...; (an instance of Transition)
+            // $workflow is a Workflow instance retrieved from the Registry (see above)
+            $title = $workflow->getMetadataStore()->getMetadata('title', $transition);
+            $this->addFlash('info', "You have successfully applied the transition with title: '$title'");
```
```suggestion
    $this->addFlash('info', "You have successfully applied the transition with title: '$title'");
```

OskarStark commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+    This is a simple example; in production you may prefer to use the :doc:`Translation </components/translation>`
+    component to manage messages in one place::
+
+    namespace App\Listener\Workflow\Task;
```
the complete php code needs to get 4 more spaces indention

pbowyer replied:
I've done this but don't understand why it's needed. Is it because the "Tip" above messes up the indentation, so code below then has to be further indented?

pbowyer replied:
No feedback so resolving.

OskarStark reviewed (changes requested):
I left a few comments

OskarStark committed:
Apply suggestions from code review

Co-Authored-By: pbowyer <peter@mapledesign.co.uk>

pbowyer committed:
Indent PHP block by an additional 4 spaces

pbowyer committed:
Change code formatting of PHP snippet per https://github.com/symfony/symfony-docs/pull/11209/files/d57fa38d903175d58d9cfbf63f20e7ced8d2fd01#r268391655

lyrixx commented on workflow/usage.rst:
```diff
@@ -450,3 +459,224 @@ The following example shows these functions in action:
+
+.. tip::
+
+    This is a simple example; in production you may prefer to use the :doc:`Translation </components/translation>`
```
IIRC, we tried to remove all occurrence of `simple` / `easy` etc word.

pbowyer replied:
Thanks @lyrixx, have reworded this sentence.

lyrixx reviewed (approved):
Except one little comment big :+1: 

Thanks you a lot for your work on this PR

pbowyer committed:
Remove the word 'simple'

pbowyer commented:
@OskarStark Your code review is showing as "Changes requested" - are you satisfied with the changes that were applied?

lyrixx commented:
@pbowyer This PR conflict. Could you try to rebase it and fix the conflict? Thanks

pbowyer commented:
@lyrixx Of course. 4.2 is the correct one to rebase onto, I believe?

lyrixx commented:
Exactly :+1: You need to rebase on top of 4.2 and update the target branch on github 
Thanks again

pbowyer commented:
@lyrixx There we go 😁

pbowyer commented:
There may be one newline missing:
```
As configured, the following property is used by the marking store::
.. code-block:: php
```
I think there should be a blank line between the two?

pbowyer committed:
Add blank line between code block and sentence above. Code block was not rendering.

OskarStark commented on workflow/usage.rst:
```diff
@@ -156,6 +165,8 @@ like this:
 
 As configured, the following property is used by the marking store::
 
+.. code-block:: php
```
This line could be removed

pbowyer replied:
Thanks @OskarStark, resolved.

OskarStark commented on workflow/usage.rst:
```diff
@@ -533,3 +544,224 @@ You can access the message from a Twig template as follows:
+
+.. versionadded:: 4.1
+
+    The transition blockers were introduced in version 4.1.
```
Please use

“... in Symfony 4.1.”

pbowyer replied:
Thanks, updated.

pbowyer committed:
Oskar's feedback

OskarStark reviewed (approved):
Thank you for your patience ✌🏻

noniagriconomie commented on workflow/usage.rst:
```diff
@@ -533,3 +542,224 @@ You can access the message from a Twig template as follows:
+                    return;
+                }
+
+                $explanation = $event->getMetadata('explanation', $event->getTransition());
```
i think you should add an `explanation` entry in the metadata of this publish transition so that a newcomer can directly spot the usage :)

noniagriconomie commented on workflow/usage.rst:
```diff
@@ -533,3 +542,224 @@ You can access the message from a Twig template as follows:
+
+.. code-block:: html+twig
+
+    <h2>Metadata</h2>
```
👍 nice examples

noniagriconomie commented:
Ref to https://github.com/symfony/symfony-docs/issues/9475

pbowyer commented:
Can we merge this please? There are more PRs for Workflow coming to fruition, and if they all get approved/merged at the same time, there will be a number of conflicts...

javiereguiluz commented:
Thanks Peter! This has been merged ... and I'll rebase the Workflow doc reorganization (#11437) with your changes.

OskarStark commented:
Great work Peter, thank you 🙏
